### PR TITLE
Specialize findprev for BitArrays

### PR DESCRIPTION
**Pull request by mbauman**
This is the first part of merging my refactor of IntSets to use BitArrays (ref. [IntSets.jl](https://github.com/mbauman/IntSets.jl) and #9679).

This adds a specialized `findprev` method for BitArrays (similar to `findnext`) that searches through the array 64 bits at a time.  Here's an exaggerated example of how much this can improve performance:

``` julia
julia> b = falses(10^10); b[10] = true;

julia> findprev(b,10^10); @time findprev(b,10^10) # Before:
elapsed time: 19.229230915 seconds (13824 bytes allocated)
10

julia> findprev(b,10^10); @time findprev(b,10^10) # After:
elapsed time: 0.265889115 seconds (96 bytes allocated)
10
```

Also adds and tests an unexported `findprevnot` (like `findnextnot`).

Secondly, in the course of writing the additional tests, I discovered that there was a bug in `findnext(::Function, ::BitArray, …)`. It did not check bounds or could return the wrong result if the function was always true or always false.  This fixes it and adds tests.

cc @carlobaldassi

### CODE REVIEW

**Comment by timholy** [comment]
This is pretty awesome. I don't have time ATM to check the code, and @carlobaldassi would be much better anyway. But the extensive test additions are obviously welcome.

File: base/bitarray.jl [hunk 1]
```diff
@@ -1377,14 +1377,90 @@ end
...
-    length(B) == 0 && return 0
-    f0 || f1 || return 0
-    f0 && f1 && return 1
     !f0 && f1 && return findnext(B, start)
-    return findnextnot(B, start)
+    f0 && !f1 && return findnextnot(B, start)
+
+    start > 0 || throw(BoundsError(B, start))
+    start > length(B) && return 0
+    length(B) == 0 && return 0
```
Comment (carlobaldassi) [inline base/bitarray.jl]: this seems redundant with the previous two checks

File: base/bitarray.jl [hunk 2]
```diff
@@ -1377,14 +1377,90 @@ end
...
 #findfirst(testf::Function, B::BitArray) = findnext(testf, B, 1)  ## defined in array.jl
 
+# returns the index of the previous non-zero element, or 0 if all zeros
+function findprev(B::BitArray, start::Integer)
+    start > 0 || return 0
+    start > length(B) && throw(BoundsError(B, start))
+
+    Bc = B.chunks
+
+    chunk_start = _div64(start)+1
```
Comment (carlobaldassi) [inline base/bitarray.jl]: should be `_div64(start-1)+1`. Test for example `findprev(trues(64), 64)`.

File: base/bitarray.jl [hunk 3]
```diff
@@ -1377,14 +1377,90 @@ end
...
+    within_chunk_start = _mod64(start)
+    mask = ~(_msk64 << within_chunk_start)
```
Comment (carlobaldassi) [inline base/bitarray.jl]: should be `mask = _msk_end(start)` I think.

File: base/bitarray.jl (hunk 2 above)
Comment (carlobaldassi) [inline base/bitarray.jl]: (You need to restore the bounds check in the following loop to catch this.)

File: base/bitarray.jl [hunk 4]
```diff
@@ -1377,14 +1377,90 @@ end
...
+
+function findprevnot(B::BitArray, start::Integer)
+    start > 0 || return 0
+    start > length(B) && throw(BoundsError(B, start))
+
+    Bc = B.chunks
+
+    chunk_start = _div64(start)+1
+    within_chunk_start = _mod64(start)
+    mask = _msk64 << within_chunk_start
```
Comment (carlobaldassi) [inline base/bitarray.jl]: I think these lines should be

```
chunk_start = _div64(start-1) + 1
mask = ~_msk_end(start)
```

File: base/bitarray.jl [hunk 5]
```diff
@@ -1377,14 +1377,90 @@ end
...
+# returns the index of the previous element for which the function returns true
+function findprev(testf::Function, B::BitArray, start::Integer)
+    f0::Bool = testf(false)
+    f1::Bool = testf(true)
+    !f0 && f1 && return findprev(B, start)
+    f0 && !f1 && return findprevnot(B, start)
+
+    start > 0 || return 0
+    start > length(B) && throw(BoundsError(B, start))
+    length(B) == 0 && return 0
```
Comment (carlobaldassi) [inline base/bitarray.jl]: same as above

**Comment by carlobaldassi** [comment]
Thanks! I left some notes, besides that all the testing is certainly most needed and welcome!
I suggest, however, when testing BitArray stuff, to test with particular fastidiousness the cases around the chunks borders, e.g. using BitArrays of length 0,1,63,64,65,127, you get the idea. These are always the trickiest to get right.

Also thanks for catching the bug in `findnext`, it was a leftover from when it was just called `findfirst` which got unnoticed in the switch, I think.

**Comment by mbauman** [comment]
Thank you for the review, @carlobaldassi!  I have addressed all comments with my updated commits.  I can rebase and squash them back into two commits again, but I think once I do that we lose your inline notes on the previous commits.

File: base/bitarray.jl (hunk 1 above)
Comment (mbauman) [inline base/bitarray.jl]: Yup, sure enough.

File: base/bitarray.jl (hunk 3 above)
Comment (mbauman) [inline base/bitarray.jl]: Ah, right you are.  That's a much more straightforward way of thinking through the logic here.  My thought process here was much more contorted.

**Comment by carlobaldassi** [comment]
Ok this looks good to me. I'm for squashing them back into two clean commits, inline comments are not really important and besides I think github keeps them anyway. Then whenever it turns green it is ok to merge for me.

**Comment by mbauman** [comment]
Rebased. Travis passed; Appveyor seems to be having some infrastructure troubles today.

**Comment by tkelman** [comment]
@carlobaldassi what about backporting just the new tests from a1289ec4fbb5c4b643fc0dd6734f10618497a829, then 37040f70a9267b7b38d59793fb01fff97c6c5e9f in its entirety? Assuming that passes locally, that is.

**Comment by mbauman** [comment]
Yeah, I should have ordered these commits the other way around.  Sorry about that.  The only relevant commits to the back port are the ones I uncomment in that commit.  I can get to this next week myself, but I'm afraid I won't have time until then.

**Comment by tkelman** [comment]
Thanks for manually doing the backport, moving the tracking label to #9833. Will look at in more detail shortly.

**Pull request merged**